### PR TITLE
Support PrintFunctionsInfo for Kubernetes Environment

### PR DESCRIPTION
**Pull request by TsuyoshiUshio**
PrintFunctionsInfo doesn't show anything if we deploy to the Kubernetes Environment. The reason is, just after the deployment, the restart of the worker process will started. So that we need to wait until the process is re-started again. 

We use the API `/admin/host/status` on Azure Functions Host that checks `processUptime` for watching the time. 
Currently, I set it as 60 sec according to the log and the timeout time of the `worker` process startup. 

I'm limited this behavior is happens only on `Kubernetes Environment`. 

# Sample output 

```bash
$ func.exe azure functionapp publish tsushihttp56
Getting site publishing info...
Creating archive for current directory...
Could not find gozip for packaging. Using DotNetZip to package. This may cause problems preserving file permissions when using in a Linux based environment.
Uploading 1.45 KB [###############################################################################]
Upload completed successfully.
Waiting for the functions host up and running ....... done
Functions in tsushihttp56:
    HttpTrigger - [httpTrigger]
        Invoke url: https://tsushihttp56.<<mydomain>>/api/httptrigger
```

CC @pragnagopa

### CODE REVIEW

File: src/Azure.Functions.Cli/Helpers/AzureHelper.cs
Comment (pragnagopa) [inline src/Azure.Functions.Cli/Helpers/AzureHelper.cs:366]: it is possible to publish a function app without any functions. We should only poll if function app is just published and if the content of the function app locally has any functions

**Comment by TsuyoshiUshio** [comment]
Hi @ahmelsayed 
I change this PR as you suggested. Could you review this? 

The change introduced.   Before Feting GetFunction(), It wait until the `/admin/host/status` response `processUptime` reaches 60000ms (I've got the time from the log  and it is also the worker startup timeout).  Then call GetFunction(). 

CC: @pragnagopa

File: src/Azure.Functions.Cli/Helpers/AzureHelper.cs
Comment (TsuyoshiUshio) [reply src/Azure.Functions.Cli/Helpers/AzureHelper.cs]: Hi @pragnagopa  
I made a change. try not to retry the GetFunctions. Sounds good?

File: src/Azure.Functions.Cli/Helpers/AzureHelper.cs
Comment (ahmelsayed) [inline src/Azure.Functions.Cli/Helpers/AzureHelper.cs:160]:
```suggestion
        internal static async Task<string> GetMasterKeyAsync(string appId, string accessToken, string managementURL)
        {
```

File: src/Azure.Functions.Cli/Helpers/AzureHelper.cs
Comment (TsuyoshiUshio) [reply src/Azure.Functions.Cli/Helpers/AzureHelper.cs:393]: Hi @ahmelsayed Thank you for the review! I try not to instantiate HttpClient every time. Is there any way to keep the static variables inside of the method? 

https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclient?view=netcore-3.1

Comment (ahmelsayed) [reply src/Azure.Functions.Cli/Helpers/AzureHelper.cs:393]: In the CLI context, the process always exists after it's done with the command, so we don't have to worry about connections leaking over time. `WaitUntilFunctionAppReadyAsync()` is called once and the process terminates. But yeah, if this was a long running service then we would want to cache the client.

File: src/Azure.Functions.Cli/Helpers/AzureHelper.cs
Comment (pragnagopa) [inline src/Azure.Functions.Cli/Helpers/AzureHelper.cs:398]: GetMasterKeyAsync can return null. Add check for null before using GetMasterKeyAsync

File: src/Azure.Functions.Cli/Helpers/AzureHelper.cs
Comment (pragnagopa) [inline src/Azure.Functions.Cli/Helpers/AzureHelper.cs:456]: Can we scope the this change to just kubernetes? Otherwise this is a change in behavior for existing apps

File: src/Azure.Functions.Cli/Helpers/AzureHelper.cs
Comment (pragnagopa) [inline src/Azure.Functions.Cli/Helpers/AzureHelper.cs:445]: do not throw exceptions as this is being retried. Only throw exception if last retry is unsuccessfull

File: src/Azure.Functions.Cli/Helpers/AzureHelper.cs
Comment (pragnagopa) [inline src/Azure.Functions.Cli/Helpers/AzureHelper.cs:433]: waiting 60secs does seem like a long delay. As Core Tools has knowledge of count of functions deployed , make this check dynamic - instead of relying on processUpTime - if host status returns 200 - make call to get function apps - if response matches list of functions in core tools - no need to wait for 60secs

**Review by pragnagopa — CHANGES_REQUESTED** [review]
Added few comments

File: src/Azure.Functions.Cli/Helpers/AzureHelper.cs
Comment (TsuyoshiUshio) [reply src/Azure.Functions.Cli/Helpers/AzureHelper.cs:393]: sure. Then, I'll include in the method.

File: src/Azure.Functions.Cli/Helpers/AzureHelper.cs
Comment (TsuyoshiUshio) [reply src/Azure.Functions.Cli/Helpers/AzureHelper.cs:433]: The `processUpTime` spec is decided when I talk with @ahmelsayed, so that I'd like to include him to this conversation.

Do you know where `count of functions deployed` ? How can I get to know the host is ready in this case? the increase of the functions deployed means host is ready?

Comment (pragnagopa) [reply src/Azure.Functions.Cli/Helpers/AzureHelper.cs:434]: As long as this change is scoped to KubeApp, this is ok for now. We can revisit improvement later.

File: src/Azure.Functions.Cli/Helpers/AzureHelper.cs
Comment (pragnagopa) [reply src/Azure.Functions.Cli/Helpers/AzureHelper.cs:446]: Will approve, once this is addressed!

File: src/Azure.Functions.Cli/Helpers/AzureHelper.cs
Comment (TsuyoshiUshio) [reply src/Azure.Functions.Cli/Helpers/AzureHelper.cs:396]: Thanks. Very good point. Done.

File: src/Azure.Functions.Cli/Helpers/AzureHelper.cs
Comment (TsuyoshiUshio) [reply src/Azure.Functions.Cli/Helpers/AzureHelper.cs]: Done the change! Thanks.

File: src/Azure.Functions.Cli/Helpers/AzureHelper.cs
Comment (TsuyoshiUshio) [reply src/Azure.Functions.Cli/Helpers/AzureHelper.cs:434]: Great! I narrow the scope!

File: src/Azure.Functions.Cli/Helpers/AzureHelper.cs
Comment (TsuyoshiUshio) [reply src/Azure.Functions.Cli/Helpers/AzureHelper.cs:446]: Hi @pragnagopa 
Could you help me to clarify this?  I see the `Retry` functions looks retry by Exception. https://github.com/Azure/azure-functions-core-tools/blob/dev/src/Azure.Functions.Cli/Helpers/RetryHelper.cs#L23
I thought that we have two opportunity to retry. 1.  status is not good. (probably functions host is not started yet) and 2. functions host is running but worker is not ready. 

In this case, How can I detect the case 1 without Exception?

File: src/Azure.Functions.Cli/Helpers/AzureHelper.cs
Comment (TsuyoshiUshio) [reply src/Azure.Functions.Cli/Helpers/AzureHelper.cs]: Done!

File: src/Azure.Functions.Cli/Helpers/AzureHelper.cs
Comment (pragnagopa) [reply src/Azure.Functions.Cli/Helpers/AzureHelper.cs:446]: Ahh thanks for clafication. Retryhelper expects an exception to be thrown and only displays if flag is set. This is OK!

File: src/Azure.Functions.Cli/Helpers/AzureHelper.cs
Comment (pragnagopa) [inline src/Azure.Functions.Cli/Helpers/AzureHelper.cs:462]:
```suggestion
            ArmArrayWrapper<FunctionInfo> functions = await GetFunctions(functionApp, accessToken, managementURL);
```

**Review by pragnagopa — APPROVED** [review]

**Pull request merged**
